src/libopensc/card-piv.c

```diff
@@ -2992,19 +3032,47 @@ static int piv_match_card(sc_card_t *card)
 	if (card->type == -1)
 		card->type = type;
 
-	card->drv_data = priv;
+	card->drv_data = priv; /* will frre if no match, or pass on to piv_init */
 	priv->aid_file = sc_file_new();
 	priv->selected_obj = -1;
 	priv->pin_preference = 0x80; /* 800-73-3 part 1, table 3 */
 	priv->logged_in = SC_PIN_STATE_UNKNOWN;
 	priv->tries_left = 10; /* will assume OK at start */
+	priv->pstate = PIV_STATE_MATCH;
 
 	/* Some objects will only be present if Histroy object says so */
 	for (i=0; i < PIV_OBJ_LAST_ENUM -1; i++)
 		if(piv_objects[i].flags & PIV_OBJECT_NOT_PRESENT)
 			priv->obj_cache[i].flags |= PIV_OBJ_CACHE_NOT_PRESENT;
 
+	sc_lock(card);
 
+	/*
+	 * detect if active AID is PIV. NIST 800-73 says Only one PIV application per card
+	 * and PIV must be the default application
+	 * This can avoid doing doing a select_aid and losing the login state on some cards
+	 * We may get interference on some cards by other drivers trying SELECT_AID before
+	 * we get to see if PIV application is still active.
+	 * putting PIV driver first might help. 
+	 * TODO could be cached too
+	 */
+	 i = piv_find_discovery(card);
```

Comment:
Are you sure that this object is present on the Yubikey? NIST Special Publication 800-73-4 says it is optional an was not present in the initial release...

Reply:
It is present on the 4 Yubico 4.2.6. Even if it is not, the the code goes ahead and tries to do a SELECT AID. 
```
3063         if (i < 0) {
3064             /* Detect by selecting applet */
3065             i = piv_find_aid(card, &aidfile);
3066         }
```

By trying the  piv_find_discovery first it avoids the Yubico issue of resetting the security state even 
when selecting the same AID. 

There are still problems in that other drivers try to select their AID before the PIV driver gets a chance to see if the PIV is the active application. So to get the full benefit of this 
OPESC_DRIVER=PIV-II is needed, and a few extra fixed by Yubico.

Reply:
OK, if it's present and does its job, fine.

Just wanted to remind you that the data from the Application Property template is still an option. Note that this is returned from SELECT and GET DATA, i.e. you don't need an extra command to retrieve the full AID on initialization, for example.

Reply:
I got a new Yubikey version 4.3.7 yesterday, and it does have a Discovery object. 

The point of using the discovery object first is it returns the active AID or fails without 
changing the security status. Using a Select AID on Yubico and some other cards causes problems even when the active application is selected again. Yubico has this problem,also see
https://github.com/makinako/OpenFIPS201/issues/1

I think we should be able to allow the user to "enter the PIN in once", to unlock a card, and use the card from other processes without having to reenter the PIN. Using disconnect=leave, and using the discovery object rather  SELECT AID will allow this. This will work with single application cards.
But not with current multiple application cards that by policy (PPIV) say if a different application on the same card is selected, the security status is post. So some of the changes are trying to address both multiple and single application cards. 


There are other places in OpenSC and code out of our control that interfere with  "enter the PIN in once". This includes other OpenSC card drivers using Select AID to test if the card supports their application. A compliant PIV card does not have that problem as NIST 800-73-3 says it ignores SELECT AID for unknown AIDS. But the Yubico does not. Other cards may have the same issue. So OpenSC when testing for a card to not cause issues for other processes or other drivers.

The use of the Discovery object tries to do the above.

Reply:
That's a misunderstanding. My suggestion was to
1. use SELECT **once** in piv_init to select the applet and to get the full Application Property template.
2. detect the PIV applet with GET DATA for the AID (4F, I believe) in **every call** of piv_card_reader_lock_obtained

Reply:
Even doing a SELECT once can cause the card to lose the login state. The discovery object '7E'
has as the AID. Its enough to show the PIV application is the active application and it does not change the login state.  If it fails, the PIV is not the active. 

The PIV does not support the '4F' It is not required. I was trying that with the script I sent last week. and none of the cards I have  support '4F'

That is why I use the '7E' discovery object that every PIV card I have has one. 
 
I use the get data of the  discovery object in every  call of piv_card_reader_lock_obtained

Since the OpenPGP has the '4F` it can use that to test. Itr will fail if PIV is active.

Reply:
OK, I see